File: 연습/부분수열의합/부분수열의합.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int n,s;
int ans; 
void go(vector<int>&a,int index,int sum){
  if(index==n) return;
  if(sum+a[index]==s) ans+=1;
  go(a,index+1,sum+a[index]);
  go(a,index+1,sum);
}

int main(){
  cin>>n>>s;
  vector<int> a(n);
  for(int i=0; i<n; i++){
    cin>>a[i];
  }
  go(a,0,0);
  cout<<ans<<'\n';
}

File: 연습/재귀/부분수열의합/부분수열의합.py
def go(a,index,sum,s):
  if index==len(a):
    if sum==s: return 1
    else: return 0
  return go(a,index+1,sum+a[index],s)+go(a,index+1,sum,s)
n,s=map(int,input().split())
a=list(map(int,input().split()))
ans = go(a,0,0,s)
if s==0 : ans-=1
print(ans)